### PR TITLE
Draft PR: Update package managers (bundler, composer, and npm_and_yarn) to support generalized abstraction

### PR DESCRIPTION
**Pull request by kbukum1**
### What are you trying to accomplish?

This PR updates the abstract package manager to include shared attributes like `ecosystem`, `name`, `version`, `deprecation_versions`, and `supported_versions`. Additionally, the individual package manager npm_and_yarn is also refactored to extend the abstract package manager and bundler and composer package managers are also updated according to the changes.

### Anything you want to highlight for special attention from reviewers?

The goal here is to centralize common functionality into the abstract `PackageManagerBase`, making each package manager more focused on its unique behavior. If there are any concerns about backward compatibility or potential edge cases in the refactoring, those are key areas to focus on.

### How will you know you've accomplished your goal?

- The package manager refactor should pass all existing tests.
- The changes should not break any current functionality, and all package managers should behave as expected with the new structure.
- Any new functionality introduced (e.g., additional attributes in the abstract manager) will have corresponding test coverage.

### Checklist

- [ ] I have run the complete test suite to ensure all tests and linters pass.
- [ ] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [ ] I have written clear and descriptive commit messages.
- [ ] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [ ] I have ensured that the code is well-documented and easy to understand.

### Commits

- update abstract package manager to include ecosystem, name, version, deprecation_versions, and supported_versions
- update bundler package manager according to the changes made in abstract package manager
- update composer package manager according to the changes made in abstract package manager
- update npm_and_yarn package manager to support generalized package manager.

### CODE REVIEW

File: bundler/lib/dependabot/bundler/package_manager.rb [hunk 1]
```diff
@@ -23,28 +24,13 @@ class PackageManager < PackageManagerBase
...
-      attr_reader :deprecated_versions
-
-      sig { override.returns(T::Array[Dependabot::Version]) }
-      attr_reader :supported_versions
-
-      sig { override.returns(T::Boolean) }
-      def unsupported?
-        # Check if the version is not supported
-        supported_versions.all? { |supported| supported > version }
+        super(
```
Comment (kbukum1) [inline bundler/lib/dependabot/bundler/package_manager.rb:27]: Moved all common fields name, version, deprecation_versions, supported_version into abstract package manager. 
Also moved `unsupported?` function into abstract package manager since bundler and composer both implements unsupported in the same way.

File: common/lib/dependabot/package_manager.rb [hunk 2]
```diff
@@ -62,12 +81,16 @@ def deprecated?
...
     #   package_manager.unsupported? #=> false
     sig { returns(T::Boolean) }
     def unsupported?
-      false
+      # If there is no defined supported_versions, we assume that all versions are supported
+      return false if supported_versions.empty?
+
+      # Check if the version is not supported
+      supported_versions.all? { |supported| supported > version }
     end
```
Comment (kbukum1) [inline common/lib/dependabot/package_manager.rb:94]: composer and bundler are implementing this method in the same way and other ecosystems are also going to apply in same way. If there is ecosystem needs to implement unsupported in a different way then it should override that in it is package manager

**Comment by kbukum1** [comment]
This PR has been split into more focused PRs:

- https://github.com/dependabot/dependabot-core/pull/10826
- https://github.com/dependabot/dependabot-core/pull/10862
